lib/openssl/ssl.rb

```diff
@@ -168,6 +166,29 @@ def verify_certificate_identity(cert, hostname)
     end
     module_function :verify_certificate_identity
 
+    def verify_hostname(san, hostname)
+      san_parts = san.split(".")
+
+      # ICANN SAC 015 recommends against wildcards on TLDs
+      return san == hostname if san_parts.size < 3
+
+      host_parts = hostname.split(".")
+      return false unless san_parts.size == host_parts.size
+      return false unless verify_domain_component(san_parts.shift, host_parts.shift)
```

Comment:
Note that verify_domain_component is only used on the leftmost domain component per RFC 6125:

```
   1.  The client SHOULD NOT attempt to match a presented identifier in
       which the wildcard character comprises a label other than the
       left-most label (e.g., do not match bar.*.example.net).
```